osu.Game/Rulesets/Objects/BarLineGenerator.cs

```diff
@@ -33,21 +31,30 @@ public BarLineGenerator(IBeatmap beatmap)
             double lastHitTime = 1 + lastObject.GetEndTime();
 
             var timingPoints = beatmap.ControlPointInfo.TimingPoints;
+            var effectPoints = beatmap.ControlPointInfo.EffectPoints;
 
             if (timingPoints.Count == 0)
                 return;
 
             for (int i = 0; i < timingPoints.Count; i++)
             {
                 TimingControlPoint currentTimingPoint = timingPoints[i];
+                EffectControlPoint? currentEffectPoint = effectPoints.FirstOrDefault(p => p.Time == currentTimingPoint.Time);
```

Comment:
it's not guaranteed that an `EffectControlPoint will be at the same point in time as the `TimingControlPoint`. Also, this is quite inefficient.

You can use `beatmap.ControlPointInfo.EffectPointAt` to avoid these two issues:

https://github.com/ppy/osu/blob/31a447fda0532fcf15f8c8251ca890c533386492/osu.Game/Beatmaps/ControlPoints/ControlPointInfo.cs#L58